linux-musl-amd64 support

petermz opened the pull request:
Hi,

This patch enables linux-musl-amd64 support. It addresses this [Alpine Linux issue](https://github.com/oracle/graal/issues/2825)

## Changes:

### mx files
* The build system calls `mx.get_os_variant()` and stores the value in a few properties around

### SVM
* At image build time, one such property is checked to tell whether we're cross compiling glibc -> musl, or run on musl natively. (This is in contrast to the initial commit in this branch, where more verbose annotation-based mechanism was used). Quite a few things in `MuslLibC` depend on that, e.g. we need `-static` for cross compiling only.
* `uint` -> `unsigned int` :)

### Sulong
* I need to pass `LIBCXX_HAS_MUSL_LIBC` to `cmake` when building `libcxx`. Here I piggyback on the existing `{{}}` mechanism, i.e. put a placeholder in `suite.py` and expand it in Python code.

### Truffle
* Do not define `_GNU_SOURCE` for musl builds.

### SDK
* The LLVM archive mentioned in `suite.py` is just the `LLVM_ORG_SRC` source package built with `-DLIBCXX_HAS_MUSL_LIBC=ON` on a musl host (Alpine 3.11)

### Installer
* I'm extending the filter language to support os-variant-arch triplet, where variant is optional. Graal updater checks the `OS_VARIANT` property and comes up with selector keys such as `linux_amd64` or `linux_musl_amd64`
* `GenerateCatalog` gets similar update, plus a new command line parameter to force variant

gilles-duboscq commented on sdk/mx.sdk/mx_sdk_vm_impl.py:
```diff
@@ -1973,8 +1976,13 @@ def _gen_gu_manifest(components, formatter, bundled=False):
+    else:
+        variant_filter = ""
     manifest["Bundle-RequireCapability"] = capability_fmt.format(os=get_graalvm_os(),
+                                                                 variant = variant_filter,
```
minor style, remove the spaces around `=` in the keyword arguments.

gilles-duboscq commented on sulong/mx.sulong/mx_sulong_suite_constituents.py:
```diff
@@ -572,7 +572,12 @@ class CMakeProject(AbstractSulongNativeProject):  # pylint: disable=too-many-anc
+        self.cmake_config = lambda: [
+            '-D{}={}'.format(k,
+                             mx_subst.path_substitutions.substitute(v)
+                                .replace('{{dollarsign}}', '$')
```
It seems a bit strange to add an additional substitution mechanism since there is already one. I know this PR is just following a trend of the local code but it might be the right time to fix it rather than extend it.
This seems to process much more than paths so it should probably use `mx_subst.string_substitutions` and new string substitutions should be registered for `$` (`<dollarsign>`) and musl (`<has_musl_libc>`?). You can look for `string_substitutions.register_no_arg` to find examples.

rschatz replied:
We fixed this already. `"\\$ORIGIN"` should now just work in `suite.py` (the `\\` escapes on the python level, resulting in the string containing `\$`, which mx interprets as literal `$`).

gilles-duboscq reviewed (commented):
Overall this looks good. See my comments.

rschatz commented on sulong/mx.sulong/suite.py:
```diff
@@ -711,10 +711,12 @@
         "LIBCXX_ENABLE_STATIC" : "NO",
         "LIBCXX_ENABLE_EXPERIMENTAL_LIBRARY" : "NO",
+        # '{{musl}}' is replaced with YES or NO in the CMakeProject.
+        "LIBCXX_HAS_MUSL_LIBC" : "{{musl}}",
```
I think this should use the `os_arch` mechanism of mx for this (see also current master, where we already use this for windows specific stuff):
```
"os": {
   "linux-musl": {
      "cmakeConfig": {
         "LIBCXX_HAS_MUSL_LIBC": "YES",
      }
   }
}
```

petermz replied:
Looks much better indeed.

gradinac commented on sdk/mx.sdk/mx_sdk_vm_impl.py:
```diff
@@ -1248,6 +1249,8 @@ def contents(self):
                 suffix = _exe_suffix
                 project_name_f = GraalVmLauncher.launcher_project_name
+                if mx.get_os() == 'linux' and mx.get_os_variant() == 'musl':
+                    build_args.append('-H:+StaticExecutableWithDynamicLibC')
```
Hmm, do we hit issues with the executables we produce without this option? If I understand right, this means that all native-image executables that we ship as part of Graal only depend on `musl` and statically link everything else

gradinac replied:
Ah, perfect - thank you! :)

gilles-duboscq commented on espresso/mx.espresso/suite.py:
```diff
@@ -290,6 +295,11 @@
                 },
+                "linux-musl": {
+                    "<others>": {
+                        "ignore": "GNU Linux-only",
```
While i understand not using the eden library on musl, it can't work to not build mokapot.
If espresso is not supported on musl, i would prefer to keep that a build failure for now.

petermz replied:
I've removed changes to this file, so building espresso now fails. I'm going to quickly look into the failure and see if I can come up with a fix in a day or two. If not, then I'll produce a separate PR later

gilles-duboscq commented on espresso/mx.espresso/suite.py:
```diff
@@ -410,6 +420,25 @@
+                            ],
+                            "LICENSE_JAVAONTRUFFLE": "file:LICENSE",
+                            "lib/": [
+                                "dependency:espresso:com.oracle.truffle.espresso.eden/<lib:eden>",
```
if eden is ignored on musl, how does that work?

petermz commented:
I've merged in the actual master branch and removed some outdated stuff. I'm now going to quickly look into the espresso failure and see if I can come up with a fix in a day or two. If not, then I'll produce a separate PR later

gilles-duboscq commented on sdk/mx.sdk/mx_sdk_vm_impl.py:
```diff
@@ -781,6 +781,7 @@ def _get_metadata(suites, parent_release_file=None):
         _metadata_dict.setdefault('JAVA_VERSION', _src_jdk.version)
         _metadata_dict.setdefault('OS_NAME', get_graalvm_os())
         _metadata_dict.setdefault('OS_ARCH', mx.get_arch())
+        _metadata_dict.setdefault('OS_VARIANT', mx.get_os_variant())
```
I am wondering if we should skip this when `mx.get_os_variant()` is an empty string

gradinac reviewed (approved):
The native-image changes look good to me

petermz commented:
I've fixed espresso build in a few steps:
* Excluded libeden since musl does not have `dlmopen` and namespaces
* Added `-Wno-error=cpp` to suppress errors related to nonstandard headers such as `<sys/poll.h>
* Added a `uint` typedef

Note that for musl builds to work, an accompanying change to mx is required. I've started a PR: https://github.com/graalvm/mx/pull/273

gilles-duboscq commented:
Thanks @petermz, the new changes for the espresso suite look good.
Note that i still don't think espresso would work with musl in this state but that's OK.

gilles-duboscq commented:
Thanks @petermz, i'll start integrating that.

petermz commented:
I've bumped mx version following integration of https://github.com/graalvm/mx/pull/273
Also `OS_VARIANT` is no longer set when variant is empty
